llpc/builder/llpcBuilderImplImage.cpp

```diff
@@ -1185,6 +1185,1185 @@ Value* BuilderImplImage::CreateImageStore(
+    bool                        isSample)       // Is sample rather than gather
+{
+    Value* pResult = nullptr;
+
+    // Extract YCbCr meta data
+    SamplerYCbCrConversionMetaData yCbCrMetaData = {};
+    yCbCrMetaData.word0.u32All = dyn_cast<ConstantInt>(CreateExtractElement(pSamplerDesc, getInt64(4)))->getZExtValue();
+    yCbCrMetaData.word1.u32All = dyn_cast<ConstantInt>(CreateExtractElement(pSamplerDesc, getInt64(5)))->getZExtValue();
+    yCbCrMetaData.word2.u32All = dyn_cast<ConstantInt>(CreateExtractElement(pSamplerDesc, getInt64(6)))->getZExtValue();
+    yCbCrMetaData.word3.u32All = dyn_cast<ConstantInt>(CreateExtractElement(pSamplerDesc, getInt64(7)))->getZExtValue();
+
+    // Only the first 4 DWORDs are sampler descriptor, we need to extract these values under any condition
+    pSamplerDesc = CreateShuffleVector(pSamplerDesc, pSamplerDesc, { 0, 1, 2, 3 });
+
+    // Init YCbCr meta data
+    uint32_t bits[3] = {yCbCrMetaData.word0.bitDepth.channelBitsR,
+                        yCbCrMetaData.word0.bitDepth.channelBitsG,
+                        yCbCrMetaData.word0.bitDepth.channelBitsB};
+
+    const int  planeNum                          = yCbCrMetaData.word1.planes;
+    const bool subsampledX                       = yCbCrMetaData.word1.xSubSampled;
+    const bool subsampledY                       = yCbCrMetaData.word1.ySubSampled;
+    const bool forceExplicitReconstruct          = yCbCrMetaData.word0.forceExplicitReconstruct;
+    const uint32_t xBitCount                     = yCbCrMetaData.word2.bitCounts.xBitCount;
+    const ComponentSwizzle swizzleR              = yCbCrMetaData.word0.componentMapping.swizzleR;
+    const ComponentSwizzle swizzleG              = yCbCrMetaData.word0.componentMapping.swizzleG;
+    const ComponentSwizzle swizzleB              = yCbCrMetaData.word0.componentMapping.swizzleB;
+    const ComponentSwizzle swizzleA              = yCbCrMetaData.word0.componentMapping.swizzleA;
+    const SamplerFilter lumaFilter               = static_cast<SamplerFilter>(yCbCrMetaData.word1.lumaFilter);
+    const SamplerFilter chromaFilter             = static_cast<SamplerFilter>(yCbCrMetaData.word1.chromaFilter);
+    const ChromaLocation xChromaOffset           = static_cast<ChromaLocation>(yCbCrMetaData.word1.xChromaOffset);
+    const ChromaLocation yChromaOffset           = static_cast<ChromaLocation>(yCbCrMetaData.word1.yChromaOffset);
+    const SamplerYCbCrRange yCbCrRange           = static_cast<SamplerYCbCrRange>(yCbCrMetaData.word0.yCbCrRange);
+    const SamplerYCbCrModelConversion yCbCrModel = static_cast<SamplerYCbCrModelConversion>(yCbCrMetaData.word0.yCbCrModel);
+
+    // Init sample descriptor for luma channel
+    Value* pSamplerDescLuma   = pSamplerDesc;
+
+    // Init sample descriptor for chroma channels
+    Value* pSamplerDescChroma = YCbCrGenerateSamplerDesc(pSamplerDesc, chromaFilter, forceExplicitReconstruct);
+
+    // Extract SQ_IMG_RSRC_WORD
```

Comment:
This code and most of what follows is 100% unmaintainable and does not take into account the differences in image descriptor definitions between different hardware generations.

Please clean this up by:
* Establish a helper method that decomposes an image descriptor into its relevant fields.
* Establish a helper method that create an image descriptor (T#) from scratch, given relevant input data.
* Rewrite this method to first decompose the image descriptor using the helper, then recreating new ones as required.

When writing the helper methods, be sure to take into account the changes in descriptor format across hardware generations.